### ISSUE TITLE
ResponseWriter should not throw RuntimeIOExceptions

### ISSUE THREAD

**Issue by ralf-g** [eclipse/jetty.project#4269]
The ResponseWriter returned by ServletResponse.getWriter() violates the contract of the PrintWriter API:
> Methods in this class never throw I/O exceptions, ...  The client may inquire as to whether any errors have occurred by invoking checkError().

Of course, ResponseWriter doesn't throw (checked) IOExceptions directly, but wraps them as unchecked RuntimeIOExceptions as soon as the application tries to write additional data. This seems to assume, that the application will use checkError() after every invocation of "write(), "print", or "append()" -- which is usually not the case. As a result applications have to add try/catch statements in order to deal with the unexpected RuntimeIOExceptions. Without such changes, ugly stacktraces are logged each time a user closes the browser during the processing of a request. Or in other words, an applicaton which runs "silently" in another servlet containers (i.e. Tomcat) must be changed for Jetty.

I understand the wish to "fail fast", but applications can do that by using the output stream (perhaps wraped by a "Writer"). If the application chooses to use a PrintWriter, then it should get a PrintWriter which behaves like a PrintWriter: I/O Errors have to be queried by "checkError().

**Comment by joakime** [comment]
This seems like a duplicate of https://bugs.eclipse.org/465754
This is reported to be fixed in Jetty 9.3.0 and newer versions of Jetty.

See commit 0dc394869b63a4573f9882275f1df59d89e2f810

Current codebase will catch all Exceptions, even `isOpen()` exceptions, and report them back to `ResponseWriter.setError(Exception)`, which allows `checkError()` to function as designed.
Do you have a reproduction case that shows otherwise?

**Comment by ralf-g** [comment]
At least in 9.4.20 the method "isOpen()" is implemented as:
```
    private void isOpen() throws IOException
    {
        if (_ioException != null)
            throw new RuntimeIOException(_ioException);

        if (_isClosed)
            throw new EofException("Stream closed");
    }

```
That means, "isOpen()" throws an "RuntimeIOException" which extends "java.lang.RuntimeException", **not** "java.io.IOException", which is indeed catched by all invokers of "isOpen()". But the "RuntimeIOException" is not catched by any of the invokers.

**Comment by joakime** [comment]
Opened PR #4290

**Comment by joakime** [comment]
Merged PR #4290
